src/rules.c

```diff
@@ -4135,6 +4141,29 @@ solver_addrecommendsrules(Solver *solv)
   solv->recommendsrules_end = solv->nrules;
 }
 
+/* add rules to exclude solvables provided by lower
+ * precedence repositories */
+void solver_addstrictrepopriorules(struct s_Solver *solv)
+{
+  Pool *pool = solv->pool;
+  Solvable *s;
+  Id sid, p, pp;
+
+  solv->strictrepopriorules = solv->nrules;
+  FOR_POOL_SOLVABLES(sid)
+  {
+	s = pool->solvables + sid;
+	int max_prio = 0;
+	FOR_PROVIDES(p, pp, s->name)
```

Comment:
we could cache the max prio here since it's the same for each `s->name` id. @mlschroe do you have a suggestion how to make a map? maybe we can allocate an array with the size of `pool->ss.nstrings`? Or is there a better way?

Reply:
E.g.

int* cache = solv_calloc(pool->ss.nstrings, sizeof(int));

then we can use that null-initialized buffer to cache values:

if (cache[s->name] == 0)
{
For provides ... 

}
else
{
max_prio = cache[s->name] - 1;
}
...